File: partition_list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a ListNode
    # @param x, an integer
    # @return a ListNode
    def partition(self, head, x):
        if head == None or head.next == None:  return head
        dummy = ListNode(-1)
        dummy.next = head
        end = dummy
        length = 0
        while end.next:
            end = end.next
            length += 1
        p = dummy
        i = 0
        while i < length:
            q = p.next
            if q.val >= x:
                end.next = q
                p.next = q.next
                end = end.next
                end.next = None
            else:
                p = p.next
            i += 1
        return dummy.next

File: Leetcode_Java/PartitionList.java
class ListNode {
    int val;
    ListNode next;
    ListNode(int x) {
        this.val = x;
    }
}

public class PartitionList {
    public ListNode partition(ListNode head, int x) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        ListNode pre = dummy;
        while (pre.next != null && pre.next.val < x) {
            pre = pre.next;
        }
        ListNode cur = pre;
        while (cur.next != null && cur.next.val >= x) {
            cur = cur.next;
            while (cur.next != null && cur.next.val < x) {
                ListNode temp = cur.next;
                cur.next = cur.next.next;
                temp.next = pre.next;
                pre.next = temp;
                pre = pre.next;
            }
        }
        return dummy.next;
    }

    public static void main(String[] args) {
        PartitionList test = new PartitionList();
        ListNode[] list = new ListNode[3];
        int[] values = {3, 1, 2};
        for (int i = 0; i < 3; i++) {
            list[i] = new ListNode(values[i]);
            if (i != 0) {
                list[i - 1].next = list[i];
            }
        }
        ListNode head = test.partition(list[0], 3);
        while (head != null) {
            System.out.print(head.val + ", ");
            head = head.next;
        }
    }
}